lib/feature/one_hot_encoder.cc

```diff
@@ -41,9 +41,9 @@ OneHotEncoderModel::transform(const int value, const int columnIndex) {
     len -= 1;
   }
 
-  SparseVector vector(len);
+  SparseVector *vector = new SparseVector(len);
```

Comment:
Is this object ever deleted in `TEST(OneHotEncoderTest, Mlir)`? 

If there is memory leak, in addition to fixing it, we probably need a systematic way to detect such memory leak to increase developer velocity and document the best practice somewhere in our documentation. 

One possible way to detect this might be to  replace `HostAllocatorType::kMalloc` with `HostAllocatorType::kLeakCheckMalloc` in `runMlirScript`. Could you try it?

It might also be useful to check whether all AsyncValues are properly de-referenced before and after a function execution in tests (and optionally in non-test setup). This logic can be found by searching for `AsyncValue::AsyncValueAllocationTrackingEnabled()` in the file `bef_executor_driver.cc`. Could you help look into this?

Reply:
I'll change all such usage into commands like below:
```suggestion
  tfrt::AsyncValueRef<SparseVector> vector =
      MakeAvailableAsyncValueRef<SparseVector>(len, &host_);
```
Such commands could help automatically delete objects through `AsyncValueRef`/`RCReference`'s internal methods. This has proved to be effective as I added a check to `SparseVector`'s destructor method to learn when it is deleted.

Changing `kMalloc` to `kLeakCheckMalloc` has no effect on the output of the tests in the case above. Given that all such objects are created through `MakeAvailableAsyncValueRef`, I wonder whether `kLeakCheckMalloc` functions correctly. So far as I can learn from TFRT, maybe only objects created through `HostContest::Construct()` can be monitored by the `HostContext`'s `HostAllocator`.

`AsyncValue::AsyncValueAllocationTrackingEnabled()` turns out to be false in the current test cases. Its documentation states that this function is only activated in debug builds, but I have not found out how to trigger this "debug" mode.

Reply:
Regarding `AsyncValueAllocationTrackingEnabled`: its source code shows that this feature is enabled when `NDEBUG` is defined. In general we can explicitly define a symbol when building a bazel target by sepcifying `--define NDEBUG=1` on the command line.

After googling this symbol, it seems that we might also be able to set the symbol by setting the c++ `compilation_mode` to `opt`. Could you try it?

Regarding `kLeakCheckMalloc`, its source code can be found at `clink_utils.cc:52` and then `profiled_allocator.cc`. It might be useful to check what it exactly checks.

Reply:
I found that by adding `-c dbg` can help invoke `NDEBUG` related operations and it helped me locate some memory related problems. Now I have fixed it and you can also have a try by running commands like follows.
```bash
bazel run //cpp_tests:feature/one_hot_encoder_test -c dbg
```
However, we cannot run `bazel test $(bazel query ...) -c dbg` yet, because the java tests would corrupt if C++ code is compiled in this debug mode. The resulting `hs_err_pidXXX.log` file indicates that the problem might be rooted from `libstdc++.so`, and I'm still looking into it. As this problem might took longer than expected, I plan to fix other comments and make a push first and keep debugging this one afterwards. In JNA codes we also created objects like `HostContext` and `AsyncValue`, so it might be improper to ignore this problem.

As for `kLeakCheckMalloc`, I read its source code and learned that it would call `exit(1)` in its destructor when it detects memory leak, and I'll use this one rather than `kMalloc` in tests.

However, I think there is no need to add tests that intentionally triggers this memory leak check. So long as tests have covered all usages of methods like load and transform and these usages have not caused memory leak, it is enough. Additional tests that try to cause memory leak would not be tests for Clink operators, but for TFRT `LeakCheckAllocator`. I also found it not simple to add such tests: `LeakCheckAllocator`'s invoking `exit(1)` has to be  caught by `EXPECT_EXIT()` in Google Test, but `EXPECT_EXIT()` only guarantees to work in a single-threaded environment([link](https://github.com/google/googletest/blob/main/docs/advanced.md#death-tests-and-threads)), while we need to execute the kernels in an environment with multiple threads.

Reply:
Yes, I agree that we don't need tests in Clink to intentionally triggers this memory leak check for the reasons you mentioned.

It will be very useful to support debugging C++ library with `dbg` when the C++ library is used via the JNA library. If it is easy to achieve right now, maybe we can just add a TODO in our meeting notes and focus on addressing other functionality stuff for now.